rust/template/distributed_datalog/src/schema.rs

```diff
@@ -0,0 +1,213 @@
+//! Schema definitions for d3log. Most of this data is to be stored in
+//! serialized form on some entity accessible from all nodes in the
+//! system.
+//!
+//! As a general note: we use BTree based maps & sets here over hashing
+//! based ones for two reasons:
+//! - they guarantee a stable order of elements which makes our
+//!   resulting computation predictable (node assignment based on data
+//!   in a hash map could yield different results on different nodes due
+//!   to randomness induced into the hashing part on purpose)
+//! - BTree based data structures can be nested nicely because `Ord` and
+//!   `PartialOrd` are defined for `BTreeMap` & `BTreeSet`, whereas hash
+//!   based ones need custom `Hash` definitions
+//!
+//! We have the following components:
+//! - Membership:
+//!   Nodes participating in the distributed computation are represented
+//!   by the `Member` struct.
+//! - Configuration:
+//!   From top to bottom we have a configuration describing the desired
+//!   state of the system, `SysCfg`. This state is essentially a
+//!   description of the required compute node configurations.
+//!   Each compute node configuration (`NodeCfg`) is comprised of a map
+//!   from relation IDs (the relations "belonging" to this node) to a
+//!   set of relation configurations, `RelCfg`. It's a set because a
+//!   relation can have multiple purposes. A relation configuration in
+//!   turn describes one property of the relation: it could have a set
+//!   of input relations and/or be fed directly from a source.
+
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::fmt::Result as FmtResult;
+use std::io::Error;
+use std::net::SocketAddr;
+use std::net::ToSocketAddrs;
+use std::option::IntoIter;
+use std::path::PathBuf;
+
+use serde::Deserialize;
+use serde::Serialize;
+
+use uuid::Uuid;
+
+use differential_datalog::program::RelId;
+
+/// An abstract representation of a node.
+///
+/// Nodes will eventually have to be assigned to members in order for a
+/// computation to happen.
+pub type Node = Uuid;
+
+/// An address of an actual member in the system.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub enum Addr {
+    /// A (ip-addr:port) pair describing how a member can be reached.
+    Ip(SocketAddr),
+}
+
+impl Display for Addr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self {
+            Addr::Ip(addr) => addr.fmt(f),
+        }
+    }
+}
+
+impl Ord for Addr {
+    fn cmp(&self, other: &Addr) -> Ordering {
+        match self {
+            Addr::Ip(addr) => match other {
+                Addr::Ip(other_addr) => addr.to_string().cmp(&other_addr.to_string()),
+            },
+        }
+    }
+}
+
+impl PartialOrd for Addr {
+    fn partial_cmp(&self, other: &Addr) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl ToSocketAddrs for Addr {
```

Comment:
We need *some* way to convert it to `SocketAddr`, because that's what we use internally right now (`TcpReceiver` & `TcpSender` are socket based). Whether that is explicitly happening through this trait impl or implicitly somewhere deep down is more a matter of taste at this point, I'd say.

I wanted a distinct `Addr` type mostly to have some more type safety (it's not just a free form `String`).
To your particular point: my intention was not to cover completely different ways of transport. I was more thinking about allowing DNS names in the future, but making these work would require at least a bit more plumbing (and matching on an `enum` makes the cases that need adjustment compiler-detectable).

I wouldn't go much further right now without a need, but let me know what you think.

Reply:
Let me merge this change. If you feel strongly I'll remove the `impl` and move it to an internal function as a follow up.

Reply:
Let me merge this change. If you feel strongly I'll remove the impl and move it to an internal function as a follow up.